### DISCUSSION TITLE
Kotlin Nullability

### DISCUSSION THREAD

**Discussion by marceloverdijk** [Netflix/dgs-framework#79] [Q&A]
Last weekend I was playing a bit with DGS codegen and Kotlin.

Imagine this schema:

```
type Person {
    id: ID!
    name: String!
    countryOfBirth: Country!
}
```

all 3 fields will not be null from a schema perspective.

This generates the following Kotlin code:

```
public data class Driver(
  public val id: String,
  public val name: String,
  public val countryOfBirth: Country
)
```

and because they were not null (`!`) in the GraphQL schema they are not null (no `?`) in the generated data class.

When a query is executed like:

```
query {
    person(id: "1") {
        name
    }
}
```

there is no need to fetch the `countryOfBirth` but the problem is the generated `Person` data class cannot be constructed without the `countryOfBirth` as it is cannot be null.

For this example it probably the country of birth would be available, but image it to be a very expensive call.

I wonder how this kind of use case can be implemented best using Kotlin and the DGS codegen?

**Comment by paulbakker** [comment]
For basic types it generates a default value (something like `data class Person(val name: String = "")`. For complex types we can't do that. I'm really not sure, other than not using `!` (which I think would be the right choice for this use case).

**Reply by marceloverdijk** [reply]
I understand that it's not possible to generate like `""` for complex types.

Not using `!` in the GraphQL schema is undesirable. I think it's not a good idea to give up on schema semantics because the implementation is Kotlin.

The only way I can see this working is that it would be possible to generate these fields with `?` in the data class.
Maybe a switch in the generator 🙈

**Reply by RIP21** [reply]
Disclaimer, I may be wrong about certain things in DGS, all the conclusions here are based off discussion above and docs that I've read so far and videos I saw :) 

@paulbakker BTW I totally in love with the approach you took here with DGS and features and that it's highly opinionated around Spring and I can't wait to move to it from `kotlin-graphql`, but the fact that we can't just generate all optional Kotlin types to partially resolve 3-4 fields at one level, and then resolve rest in a separate field level resolver methods is a bit annoying :)

And, related to not using `!`, I being mostly a Frontend dev (ex Java dev) with Typescript being my daily driver for the last 4 years, simply cannot understand why all of the examples you folks are showing are not using `!` at all.
It will be total hell on earth for me to develop and work with such GQL API on client-side code as all typings will be generated from GraphQL Schema hence then all fields will require null checks, all arguments are optional meaning benefits of using TS going below zero literally :) Aside from field name autocomplete + typo protection :)
I'm unsure how Frontend devs at Netflix can be happy with such schemas if they're using TS and [graphql-codegen](https://www.graphql-code-generator.com/).

One of the benefits of GraphQL for me is that it's basically strictly typed contract all the way, from DB (to some extent) to Backend, to clients. If anything changes at the very bottom it will all be reflected all the way up to the client making it mega-powerful in combo with TS or any other typed language as a client of GraphQL.
Surely as more strict, it's as harder it to evolve, but there are deprecated fields and other strategies to evolve API painfully for all the consumers :)

Having a switch in the generator to generate all optional data types would be nice :) @marceloverdijk if I'll make eventually some generator based on [graphql-codegen](https://www.graphql-code-generator.com/) that will do exactly that will let you know :P I plan a migration to a next week or so, so maybe then :)

Hope it all makes sense :)

**Reply by paulbakker** [reply]
A switch to just always use optional would be ok I think. There is a lot of opinions woven into such things, so it would be useful to have that configurable.

Why would you consider creating a generator based on graphql-codegen though? Not just for this feature, I assume?

**Reply by RIP21** [reply]
@paulbakker just for this feature :)
If that would be implemented then there is no need.
I still haven't tried DGS even on "hello world" example so I'm unsure what DGS generator generates exactly and how it helps :) So maybe I'll come up with some other ideas as I'm heavy user of graphql-codegen but for Typescript server side and client side generators.

**Reply by marceloverdijk** [reply]
Yes, as I said before a generator switch to use optional `?` would be a good improvement for this use case.
It should not be hard to this; @RIP21 interested in providing a PR? Although I'm not using Kotlin atm for my DGS server, I could have a go at this PR if you cannot do it yourself.

**Reply by RIP21** [reply]
@marceloverdijk I'm more into JS ecosystem :) So I would rather hack something quickly using TS :) But Paul is already did what we wanted below, so :)

**Comment by paulbakker** [comment]
I have a PR with a new `kotlinAllFieldsOptional` setting. When set to true, it always generates nullable fields for Kotlin.

That being said, I also started to re-consider what to do for complex types, even when the new setting is `false`. Since we can't set a default value for complex types, should we just always make fields of complex types non-nullable?

**Reply by RIP21** [reply]
Thanks @paulbakker for the quick execution :)

I don't know what exactly DGS generates, but I guess DTOs that are 1:1 with what is in the schema and with setting `false` they reflect truth 1:1, while with `true` all the fields become optional hence with a default value of `null`.

I still had time to check its actual output of the DGS generator and never had a change to get my hands on it properly, so sorry for my stupidity/unawareness :P 

So I guess @marceloverdijk maybe you can input something here?

**Reply by marceloverdijk** [reply]
Personally I would remove **all** the default values. Also for simple types like String.

Imo when having a required String type - same for complex types - I thing the user should always set it explicitly.

**Reply by paulbakker** [reply]
I have described the new flag a little better in the PR: https://github.com/Netflix/dgs-codegen/pull/96

@marceloverdijk could you take a look if that's the desired behavior?

**Reply by RIP21** [reply]
@paulbakker Looks good for me :) 

This means that I can resolve partially in the root resolver (field on the Query type) that will have this as a return type with whatever I have and then resolve the rest in the field level resolvers. Which is exactly what needed :) 

It will require testing to avoid schema definition to implementation/runtime missmatching, but it's a good tradeoff for me.
Although if we will be able to statically check what is set on the root resolver and then make sure that field-level resolvers are also written hence all the requirements for the type are there, will be awesome, but unsure it's possible without AST parsing of Kotlin/Java code by some CLI tool :)
Also, some sort of IDE Plugin would be awesome for outstanding DX :) 
Anyway, just thoughts to the air. Thank you for the change! :)

If something fragile means will be better tested :P So means it's not that bad of a side effect :D

**Reply by marceloverdijk** [reply]
Yes it looks good! 💪

**Reply by paulbakker** [reply]
Thanks for the feedback, will merge the PR! :)

**Reply by RIP21** [reply]
@marceloverdijk you can also set this to be "answered" :P

**Reply by jmbarroso** [reply]
@paulbakker found this thread, and it was something we needed. Would it be worth adding it to the documentation?

**Comment by moreginger** [comment]
I also just ran into this. Allowing null isn't so bad on the server side, but in our Typescript UI the APIs generated from the same schema then need extra null checks / coercion to compile. On the other hand, it's a bit of a shame to have to use `kotlinAllFieldsOptional` (though better than breaking the schema).

Could you generate a sealed class for each complex object with a default "NoFoo" variant vs the real Foo, or does that get messy in other ways?

**Comment by srinivasankavitha** [comment]
We just merged a new feature in dgs-codegen with the ability to
generateKotlinNullableClasses - `config.generateKotlinNullableClasses` as
part of this PR: https://github.com/Netflix/dgs-codegen/pull/359. That
might be what you are looking for.

**Reply by moreginger** [reply]
Thanks for the info. I'm not quite sure - I can't see an example that covers my case. I'll keep an eye on this feature when it drops :)
